# Exercício
## Ensembles e K-nearest neighbors
### Alunos (Nome e número usp):
- Fernanda Tostes Marana (4471070)
- Matheus Aparecido do Carmo Alves (9791114)
- Thais Bianchini (9791010)
 
---

Para esse exercício, vamos utilizar novamente o dataset [Breast Cancer Winsconsin](https://scikit-learn.org/stable/datasets/index.html#breast-cancer-dataset).



---

### Questão 1.

Carregue o dataset Breast Cancer do módulo `sklearn.datasets` e normalize os dados.

In [0]:
import numpy as np
import pandas as pd
from sklearn import datasets
from sklearn.preprocessing import StandardScaler

# 1. Loading breast cancer dataset
breast_cancer = datasets.load_breast_cancer()
X = breast_cancer['data']
y = breast_cancer['target']

# 2. Normalising the data
# - here we are standardising features by removing the mean and scaling to unit 
# variance, i.e.: z = (x-u)/s, where x is the data vector, u is the mean and
# s is the standard deviation 
scaler = StandardScaler()
X = scaler.fit_transform(X,y)

# 3. Presenting the normalised dataset
data = np.column_stack((X,y))
feature_names = list(breast_cancer['feature_names'])
feature_names.append('class')
breast_cancer_df = pd.DataFrame(data,columns=feature_names)
breast_cancer_df

mean radius  mean texture  ...  worst fractal dimension  class
0       1.097064     -2.073335  ...                 1.937015    0.0
1       1.829821     -0.353632  ...                 0.281190    0.0
2       1.579888      0.456187  ...                 0.201391    0.0
3      -0.768909      0.253732  ...                 4.935010    0.0
4       1.750297     -1.151816  ...                -0.397100    0.0
..           ...           ...  ...                      ...    ...
564     2.110995      0.721473  ...                -0.709091    0.0
565     1.704854      2.085134  ...                -0.973978    0.0
566     0.702284      2.045574  ...                -0.318409    0.0
567     1.838341      2.336457  ...                 2.219635    0.0
568    -1.808401      1.221792  ...                -0.751207    1.0

[569 rows x 31 columns]



---

### Questão 2.

Aqui definiremos funções que serão utilizadas para simplificar as operações que faremos posteriormente. Dessa forma, implemente as funções abaixo:

*   A função `get_mean_accuracy(model, X, y)` recebe um modelo `model` e um conjunto de atributos `X` e labels `y`. Ela deve calcular a acurácia do modelo utilizando 10-fold cross-validation estratificado e retornar a acurácia média dos 10 folds
*   A função `evaluate_models(models, X, y)` recebe um conjunto de modelos definidos por um dicionário e exibe, para cada modelo, seu nome seguido da sua acurácia (calculada com a função `get_mean_accuracy`). Um exemplo de saída para o dicionário `exemplo` abaixo seria:
> A acurácia do modelo "Knn (n_neighbors = 5)" é 85.00%
>
> A acurácia do modelo "DT (gini)" é 80.00%

*   Finalmente, implemente a função `create_name_list(models)`. Essa função deve receber um dicionário e retornar uma lista contendo, para cada elemento do dicionário, uma tupla (chave, valor). Para o dicionário `exemplo`, essa função retornaria:
> `[ ('Knn (n_neighbors = 5)', KNeighborsClassifier(n_neighbors=5)), ('DT (gini)', DecisionTreeClassifier(criterion="gini"))]`


```
exemplo = {
    'Knn (n_neighbors = 5)': KNeighborsClassifier(n_neighbors=5),
    'DT (gini)': DecisionTreeClassifier(criterion="gini"),
}
```



In [0]:
from numpy import mean
from sklearn.model_selection import cross_validate

def get_mean_accuracy(model, X, y):
  # Applying the stratified 10-fold cross-valitation
  metrics = cross_validate(model, X, y, cv=10)

  # Returning the mean accuracy score
  return(mean(metrics['test_score'])) 

def evaluate_models(models, X, y):
  for (model_name,model) in models:
    print('A acurácia do modelo \"'+str(model_name)+'\" é '+\
          str(np.round(100*get_mean_accuracy(model,X,y),4))+'%')

def create_name_list(models):
  list1 = [(k, v) for k, v in models.items()]
  return (list1)



---

### Questão 3.

Defina modelos de SVM (`sklearn.svm.SVC`) e MLP (`sklearn.neural_network.MLPClassifier`) para servir de *baseline* de comparação para modelos futuros. Para isso, teste ao menos 3 configurações de SVM e 3 configurações de MLP. Sua baseline será a melhor acurácia (utilizando 10-fold cross-validation estratificado) entre essas configurações. Utilize as funções definidas na Questão 2 quando necessário.

In [0]:
from sklearn.neural_network import MLPClassifier
from sklearn.svm import SVC

experiments = { 
    "SVM Linear" : SVC(kernel='poly',degree=1,probability=True),
    "SVM 3-Poly" : SVC(kernel='poly',degree=3,probability=True),
    "SVM 9-Poly" : SVC(kernel='poly',degree=9,probability=True),
    "MLP 3-HL" : MLPClassifier(hidden_layer_sizes=(30,30,30,),solver='lbfgs'),
    "MLP 6-HL" : MLPClassifier(hidden_layer_sizes=(30,30,30,30,30,30,),solver='lbfgs'),  
    "MLP 9-HL" : MLPClassifier(hidden_layer_sizes=(30,30,30,30,30,30,30,30,30,),solver='lbfgs'),
}
list_ready = create_name_list(experiments)
evaluate_models(list_ready,X,y)

A acurácia do modelo "SVM Linear" é 97.713%
A acurácia do modelo "SVM 3-Poly" é 90.3383%
A acurácia do modelo "SVM 9-Poly" é 79.2638%
A acurácia do modelo "MLP 3-HL" é 97.3622%
A acurácia do modelo "MLP 6-HL" é 95.4323%
A acurácia do modelo "MLP 9-HL" é 90.1598%




---

### Questão 4.

Agora defina dois dicionários de classificadores. Um dicionário deve conter apenas classificadores KNN (`sklearn.neighbors.KNeighborsClassifier`), enquanto o outro deve conter apenas classificadores DT (`sklearn.tree.DecisionTreeClassifier`). Crie ao menos 3 configurações diferentes para cada tipo de classificador. Depois calcule e exiba a acurácia de cada configuração criada utilizando 10-fold cross-validation estratificado.

In [0]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier

experiments_knn = { 
    "KNN 10-NN" : KNeighborsClassifier(n_neighbors=10),
    "KNN 25-NN" : KNeighborsClassifier(n_neighbors=25),
    "KNN 50-NN" : KNeighborsClassifier(n_neighbors=50),
}
experiments_dt = { 
    "DTC 10-MD (gini)" : DecisionTreeClassifier(criterion='gini',max_depth=10),
    "DTC 10-MD (entropy)" : DecisionTreeClassifier(criterion='entropy',max_depth=10),
    "DTC 25-MD (gini)" : DecisionTreeClassifier(criterion='gini',max_depth=25),
    "DTC 25-MD (entropy)" : DecisionTreeClassifier(criterion='entropy',max_depth=25),
    "DTC 50-MD (gini)" : DecisionTreeClassifier(criterion='gini',max_depth=50),
    "DTC 50-MD (entropy)" : DecisionTreeClassifier(criterion='entropy',max_depth=50),
}

# Testing knn models
list_ready = create_name_list(experiments_knn)
evaluate_models(list_ready,X,y)

# Testing dt models
list_ready = create_name_list(experiments_dt)
evaluate_models(list_ready,X,y)

A acurácia do modelo "KNN 10-NN" é 97.0113%
A acurácia do modelo "KNN 25-NN" é 95.6078%
A acurácia do modelo "KNN 50-NN" é 95.4323%
A acurácia do modelo "DTC 10-MD (gini)" é 90.8741%
A acurácia do modelo "DTC 10-MD (entropy)" é 92.7882%
A acurácia do modelo "DTC 25-MD (gini)" é 91.2187%
A acurácia do modelo "DTC 25-MD (entropy)" é 92.7851%
A acurácia do modelo "DTC 50-MD (gini)" é 91.3972%
A acurácia do modelo "DTC 50-MD (entropy)" é 92.9637%




---

### Questão 5.

Agora você deve definir um dicionário contendo diferentes configurações de ensembles a serem testados. Utilize as classes `StackingClassifier`, `AdaBoostClassifier` e `VotingClassifier` do módulo `sklearn.ensemble` (fique a vontade para escolher outros tipos de ensembles desse mesmo módulo). Crie pelo menos 5 configurações diferentes tomando como estimadores os modelos definidos na Questão 4. Calcule a acurácia de cada configuração de ensemble. 

Obs.: Lembre-se de utilizar as funções definidas anteriormente

In [0]:
from sklearn.ensemble import StackingClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.ensemble import VotingClassifier
from sklearn.ensemble import RandomForestClassifier

estimators = [('SVM Linear',experiments['SVM Linear']),\
              ('MLP 3-HL',experiments['MLP 3-HL']),\
              ('KNN 10-NN',experiments_knn['KNN 10-NN']),\
              ('DTC 10-MD (entropy)',experiments_dt['DTC 10-MD (entropy)'])]

experiments_ens = { 
    "SC PP" : StackingClassifier(estimators,cv=10,stack_method='predict_proba'),
    "SC P " : StackingClassifier(estimators,cv=10,stack_method='predict'),
    "ABC 1.0-LR" : AdaBoostClassifier( n_estimators=50, learning_rate=1.0),
    "ABC 0.7-LR" : AdaBoostClassifier( n_estimators=50, learning_rate=0.7),
    "ABC 0.5-LR" : AdaBoostClassifier( n_estimators=50, learning_rate=0.5),
    "VC Soft" : VotingClassifier(estimators,voting='soft'),
    "VC Hard" : VotingClassifier(estimators,voting='hard'),
}

list_ready = create_name_list(experiments_ens)
evaluate_models(list_ready,X,y)

A acurácia do modelo "SC PP" é 98.4148%
A acurácia do modelo "SC P " é 97.5376%
A acurácia do modelo "ABC 1.0-LR" é 96.131%
A acurácia do modelo "ABC 0.7-LR" é 96.6604%
A acurácia do modelo "ABC 0.5-LR" é 95.9586%
A acurácia do modelo "VC Soft" é 97.5376%
A acurácia do modelo "VC Hard" é 97.005%




---

### Questão 6. 

Discuta as acurácias obtidas nas Questões 3, 4 e 5. Houve algum ganho ao utilizar Ensembles? Em qual situação houve maior ganho?

**RESPOSTA:**

Resguardando as acurácias obtidas nas Questões 3, 4 e 5, temos que as melhores classificações de cada modelo foram:

* **SVM Linear :** 97.713%
* **MLP 3-HL :** 95.0815%
* **KNN 10-NN :** 97.0113%
* **DTC 10-MD (entropy):** 93.6779%

Estes melhores resultados guiaram a decisão do conjunto de estimadores a serem acoplados nos Ensembles.

Ao se utilizar estes métodos sozinhos, como classificadores (considerando as arquiteturas definidas neste exercício), alcançou-se a melhor acurácia com a aplicação da SVM Linear sobre o problema (97.713% de acurácia).

A partir da aplicação de Ensembles, foi possível aprimorar esse resultado significativamente, alcançando uma acúracia maior que 98% na utilização do método de Stacking como Ensemble e classificador final (arquitetura "SC PP"), podendo variar um pouco devido a natureza de aleatorização do algoritmo.
Esta melhora era esperada, dada a natureza e funcionamento de Ensembles.

De maneira direta, todos os Ensambles garantiram uma melhora na classificação final para algum dos classificadores utilizados. O que mais se aprimorou foi o modelo DTC 10-MD que antes apresentava um menor taxa de acurácia em seus resultados. Adicionalmente, a situação que garantiu o maior ganho foi a utilização da arquitetura "SC PP" (como comentado anteriormente).

Desta forma, pode-se observar que existe ganho na aplicação de Ensembles, sendo necessário somente estudá-los e modelá-los para melhorar o resultado para o problema como um todo.
